sbcl: cleanup old versions

hraban opened the pull request:
###### Description of changes

Rebased #203659 onto master because this doesn't actually rebuild much. See that PR for details.

~This PR depends on #209023.~

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

hraban committed:
pgloader: use latest SBCL, not pinned to 2.2.6

7c6f434c commented:
@ofborg build lispPackages.stumpwm stumpwm

7c6f434c commented:
Do we have an attribute path with working StumpWM after pruning? Maybe point top-level stumpwm attribute there…

(Or maybe updating Quicklisp stuff now will be compatible with fresh SBCL, it should be but I haven't got around to it)

7c6f434c commented:
Hmmm I hoped to merge #209405 and then this…

hraban commented:
Thanks for staying on top of it @7c6f434c .

7c6f434c commented:
I think _now_ StumpWM uses fresh SBCL. On the bright side, this is a side effect of a large refactoring (coming together with an update). On the dark side, it created a few merge conflicts.

But now I think that indeed dropping almost all the old version is fine. Could you please rebase (or redo, given this is mainly deletions) this?

On another note, do you want to be in the Lisp team (another part of the refactors…)? Or maybe added just to the list of SBCL maintainers?

hraban commented:
Rebased and pushed.

> On another note, do you want to be in the Lisp team (another part of the refactors…)? Or maybe added just to the list of SBCL maintainers?

Sure, sounds fun. 👍 thanks for the invite.

7c6f434c commented:
CLPM seems to refer to SBCL 2.0.8 by attribute name. Does it work with 2.3.0?

And I guess you can also add yourself to the team in `maintainers/team-list.nix` if the PR needs an update anyway…

7c6f434c commented:
(Hmm, to add you to the GitHub team I need you to be in the GitHub NixOS org, and once you are in the maintainer list for SBCL via the team, I hope you get an invite from a periodic automated invite-new-maintainers run)

hraban committed:
Add new maintainer: hraban

hraban committed:
sbcl: remove old versions, leave 2.1.9 for CLPM

hraban committed:
clpm: do fixup, just don’t strip

Less invasive change to the default build process, with the same result.

nagy commented on pkgs/development/tools/clpm/default.nix:
```diff
@@ -4,7 +4,9 @@
-, sbcl_2_0_8
+# Once that bug is fixed, replace this with regular ‘sbcl’ and remove all
+# references to sbcl 2.1.9 from nixpkgs, including from sbcl/2.x.nix.
+, _sbcl_2_1_9
```
If this is the only package left to rely on another sbcl version, it could be justified to write something like this:
```nix
let
  # Broken on newer versions:
  # https://gitlab.common-lisp.net/clpm/clpm/-/issues/51
  sbcl = sbcl.overrideAttrs (_: {
    src = fetchurl {
      url = "mirror://sourceforge/project/sbcl/sbcl/2.1.9/sbcl-2.1.9-source.tar.bz2";
      sha256 = "189gjqzdz10xh3ybiy4ch1r98bsmkcb4hpnrmggd4y2g5kqnyx4y";
    };
  });
in ...
```
This could allow us to remove the rest of this version mapping already without having to wait for that bug to be fixed and released.

hraban replied:
How about this? Because the `wrapLisp` function itself doesn't provide an .override, you can directly access the `.override` provided by callpackage on the underlying sbcl call to 2.x.nix. Long story short: you can just import 2.1.9 directly in clpm, from 2.x.nix, without going through all-packages. This gets us half-way there.

nagy replied:
Sure, this looks like an improvement to me.

hraban committed:
refactor: don’t expose sbcl 2.1.9 to all-packages

Only give access to it when explicitly requested by CLPM. This prevents
polluting all-packages namespace with a deprecated package.

storvik commented:
I would be fine with removing CLPM from nixpkgs. It has not been actively developed past year and I doubt anyone is using it. Also, as you may have noticed, it usually breaks when updating SBCL. 

Just a suggestion, not sure what's the policy regarding removing a package from nixpkgs.

7c6f434c commented:
@storvik Hmm, the latest snapshot from development branch might fix the current SBCL compatibility issue (it is Esrap-induced, not «internal»), but if you as the maintainer find the current latest upstream snapshot not good enough for considering useful-frozen state, I think you can just drop the package with little ceremony. 

I think there are many Lisp packages that are useful with no upstream changes, so as you find the current upstream state not useful as is, please state this too in the removal message.

(Feel free to @-mention me for a not-immediate but bounded-delay merge)

(Checking what is our special-needs SBCL version story after the refactoring is a useful thing in itself, in my opinion…)
